File: Tree/235_LowestCommonAncestorOfBinarySearchTree.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        min_val = min(p.val, q.val)
        max_val = max(p.val, q.val)
        while root:
            value = root.val
            if min_val <= value <= max_val:
                return root
            elif max_val < value:
                root = root.left
            else:
                root = root.right
        return None

File: Tree/235_LowestCommonAncestorOfBinarySearchTree.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-09-20 19:37:23
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // Easy to understand
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int min = p->val < q->val ? p->val : q->val;
        int max = p->val < q->val ? q->val : p->val;
        while(root){
            if(root->val < min){
                root = root->right;
            }
            else if(root->val > max){
                root = root->left;
            }
            else{
                return root;
            }
        }
        return nullptr;
    }
};
